src/lib/crypto/kimchi_backend/gadgets/foreign_field.ml

```diff
@@ -1127,6 +1127,51 @@ let mul (type f) (module Circuit : Snark_intf.Run with type field = f)
         } ) ;
   Element.Standard.of_limbs (remainder0, remainder1, remainder2)
 
+(* Gadget to constrain conversion of bytes array (output of Keccak gadget)
+   into foreign field element with standard limbs (input of ECDSA gadget) *)
+let _keccak_to_ecdsa (type f)
+    (module Circuit : Snark_intf.Run with type field = f)
+    (bytestring : Circuit.Field.t array) (fmod : f standard_limbs)
+    (fmod_bitlen : int) =
+  let open Circuit in
+  (* C1: Check modulus_bit_length = # of bits you unpack 
+   * This is partly implicit in the circuit given the number of byte outputs of Keccak:
+   * · input_bitlen < fmod_bitlen : OK
+   * · input_bitlen = fmod_bitlen : OK
+   * · input_bitlen > fmod_bitlen : CONSTRAIN
+   * Check that the most significant byte of the input is less than 2^(fmod_bitlen % 8)
+   *)
+  let input_bitlen = Array.length bytestring * 8 in
+  if input_bitlen > fmod_bitlen then
+    (* For the most significant one, constrain that it is less bits than required *)
+    Lookup.less_than_bits (module Circuit) (fmod_bitlen % 8) bytestring.(0) ;
```

Comment:
I mean, I guess this is part of the integration tests we want to do. It could be anything, I don't know what's the standard in other ECDSA implementations. We should compare this when we know more, in any case good to leave this open.